src/motion/space_groups.F

```diff
@@ -0,0 +1,1229 @@
+         spgr%n_core = n_shell
+      ELSE
+         spgr%n_core = 0
+      ENDIF
+
+      IF (PRESENT(n_shell)) THEN
+         spgr%n_shell = n_shell
+      ELSE IF (PRESENT(n_core)) THEN
+         spgr%n_shell = n_core
+      ELSE
+         spgr%n_shell = 0
+      ENDIF
+
+      IF ( .NOT.( spgr%nparticle == (spgr%n_atom + spgr%n_shell) ) ) THEN
+         CPABORT("spgr_gen: nparticle not equal to natom + nshell.")
+      END IF
+
+      spgr%nparticle_sym = spgr%nparticle
+      spgr%n_atom_sym = spgr%n_atom
+      spgr%n_core_sym = spgr%n_core
+      spgr%n_shell_sym = spgr%n_shell
+
+      ! output unit
+      IF (PRESENT(iounit)) THEN
+         spgr%punit = iounit
+      ELSE
+         spgr%punit = 6
+      END IF
+
+      ! accuracy for symmetry
+      IF (PRESENT(eps_geo)) THEN
+         spgr%eps_geo = eps_geo
+      ELSE
+         spgr%eps_geo = 1.e-4_dp
+      END IF
+
+      ! accuracy for symmetry
+      IF (PRESENT(eps_sym)) THEN
+         spgr%eps_sym = eps_sym
+      ELSE
+         spgr%eps_sym = 1.e-4_dp
+      END IF
+
+      ! vector to test reduced symmetry
+      IF (PRESENT(pol)) THEN
+         spgr%pol(1) = pol(1)
+         spgr%pol(2) = pol(2)
+         spgr%pol(3) = pol(3)
+      ELSE
+         spgr%pol = 0.0_dp
+      END IF
+
+      ALLOCATE(spgr%lat(spgr%nparticle))
+      spgr%lat=.TRUE.
+
+      IF (PRESENT(ranges)) THEN
+        n_sr_rep = SIZE(ranges, 2)
+        DO i = 1, n_sr_rep
+          DO j = ranges(1, i), ranges(2, i)
+            spgr%lat(j)=.FALSE.
+            spgr%nparticle_sym=spgr%nparticle_sym-1
+            IF (j <= spgr%n_atom) THEN
+               spgr%n_atom_sym = spgr%n_atom_sym -1
+            ELSE IF (j > spgr%n_atom .AND. j <= spgr%nparticle) THEN
+               spgr%n_core_sym = spgr%n_core_sym -1
+               spgr%n_shell_sym = spgr%n_shell_sym -1
+            ELSE
+               CPABORT("Symmetry exclusion range larger than actual number of particles.")
+            ENDIF
+          END DO
+        END DO
+      END IF
+
+      ALLOCATE(tmp_coor(3, spgr%n_atom_sym), tmp_types(spgr%n_atom_sym))
+
+      j = 0
+      DO i = 1, spgr%n_atom
+        IF (spgr%lat(i)) THEN
+          j = j+1
+          tmp_coor(:, j) = scoor(:, i)
+          tmp_types(j) = types(i)
+        END IF
+      END DO
+
+      !..set cell values
+      NULLIFY(spgr%cell_ref)
+      CALL cell_create(spgr%cell_ref)
+      CALL cell_copy(cell, spgr%cell_ref)
+      SELECT CASE (gopt_env%type_id)
+      CASE (default_minimization_method_id, default_ts_method_id)
+         CALL init_cell(spgr%cell_ref, hmat=cell%hmat)
+      CASE (default_cell_method_id)
+         SELECT CASE (gopt_env%cell_method_id)
+         CASE (default_cell_direct_id)
+            CALL init_cell(spgr%cell_ref, hmat=gopt_env%h_ref)
+         CASE (default_cell_geo_opt_id, default_cell_md_id)
+            CPABORT("SPACE_GROUP_SYMMETRY SHOULD NOT BE INVOKED DURING THE CELL STEP.")
+         CASE DEFAULT
+            CPABORT("")
```

Comment:
no blank `CPABORT` if possible, since they always provide some sort of documentation as well